flexidag/src/blockdag.rs

```diff
@@ -150,20 +156,26 @@ impl BlockDAG {
         )?;
 
         // Update reachability store
-        let mut reachability_store = self.storage.reachability_store.clone();
+        let reachability_store = self.storage.reachability_store.clone();
         let mut merge_set = ghostdata
             .unordered_mergeset_without_selected_parent()
-            .filter(|hash| self.storage.reachability_store.has(*hash).unwrap());
-        match inquirer::add_block(
-            &mut reachability_store,
+            .filter(|hash| self.storage.reachability_store.read().has(*hash).unwrap())
+            .collect::<Vec<_>>()
+            .into_iter();
+        let mut reachability_writer = reachability_store.write();
+        let add_block_result = inquirer::add_block(
+            reachability_writer.deref_mut(),
             header.id(),
             ghostdata.selected_parent,
             &mut merge_set,
-        ) {
+        );
+        drop(reachability_writer);
```

Comment:
please see this https://amanieu.github.io/parking_lot/parking_lot/struct.RwLock.html
there is a demo
```rust
use parking_lot::RwLock;

let lock = RwLock::new(5);

// many reader locks can be held at once
{
    let r1 = lock.read();
    let r2 = lock.read();
    assert_eq!(*r1, 5);
    assert_eq!(*r2, 5);
} // read locks are dropped at this point

// only one write lock may be held, however
{
    let mut w = lock.write();
    *w += 1;
    assert_eq!(*w, 6);
} // write lock is dropped here
```